src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt

```diff
@@ -100,28 +102,45 @@ open class SmkFileReference(
         return null
     }
 
-
-    private fun findVirtualFile(): VirtualFile? {
+    private fun findVirtualFile(searchFromRoot: Boolean = false): VirtualFile? {
         val stringLiteral = element.argumentList?.arguments?.firstOrNull {
             it is PyStringLiteralExpression && it.stringValue == path
         } ?: return null
 
-        val parentFolder = element.containingFile.originalFile.virtualFile?.parent ?: return null
-
-        val filePath = (stringLiteral as? PyStringLiteralExpression)?.stringValue
+        val filePath = (stringLiteral as? PyStringLiteralExpression)?.stringValue!!
```

Comment:
Not clear how it works `(stringLiteral as? PyStringLiteralExpression)?` is marked as nullable, returns `null` or expression, but than is declared as not-null `!!`

Reply:
Moreover prev line always returns `PyStringLiteralExpression` with `stringValue == path`, so there is no need to calculate `stringValue` again. Type checker fails to infer that stringLiteral is `PyStringLiteralExpression` from the above code, but we could refactor it using:

```kotline
        val stringLiteral = element.argumentList?.arguments?.firstOrNull { 
                it is PyStringLiteralExpression &&  it.stringValue == path 
            } as PyStringLiteralExpression? ?: return null

        // or
        val stringLiteral = element.argumentList?.arguments
            ?.filterIsInstance(PyStringLiteralExpression::class.java)
            ?.firstOrNull { it.stringValue == path }
            ?: return null
        
// in both cases `stingLiteral` type will be inferred as `PyStringLiteralExpression`
            
```

Reply:
Also, is not clear why we need `stringLiteral` check in the first line of the method, seems could be refactored just to:
```kotlin
    private fun findVirtualFile(searchFromRoot: Boolean = false): VirtualFile? {
        // Try to find using relative path
        val relativeFile = if (searchFromRoot) {
            //search in all content roots
            ProjectRootManager.getInstance(element.project).contentRoots.firstOrNull { root ->
                root.findFileByRelativePath(path) != null
            }
        } else {
            element.containingFile.originalFile.virtualFile?.parent
                ?.findFileByRelativePath(path)
        }
        if (relativeFile != null) {
            return relativeFile
        }
        // Trying to find the file anywhere
        val vfm = VirtualFileManager.getInstance()
        val localFS = vfm.getFileSystem(LocalFileSystem.PROTOCOL)
        return localFS.findFileByPath(path) ?: vfm.findFileByUrl(path)
    }
```

Reply:
Ok, this is really better, but we have to use firstNotNullOfOrNull